contracts/core/src/token/OmniGasPump.sol

```diff
@@ -0,0 +1,225 @@
+import { OmniGasStation } from "./OmniGasStation.sol";
+
+/**
+ * @title OmniGasPump
+ * @notice A unidirectional cross-chain gas exchange, allowing users to swap native ETH for native OMNI.
+ */
+contract OmniGasPump is XAppUpgradeable, OwnableUpgradeable, PausableUpgradeable {
+    /// @notice Emitted when the max swap is set
+    event MaxSwapSet(uint256 max);
+
+    /// @notice Emitted when the gas station is set
+    event GasStationSet(address station);
+
+    /// @notice Emitted when the toll is set
+    event TollSet(uint256 pct);
+
+    /**
+     * @notice Emitted on a fillUp
+     * @param recipient Address on Omni to send OMNI to
+     * @param owed      Total amount owed to `recipient` (sum of historical fillUps()), denominated in OMNI.
+     * @param amtETH    Amount of ETH paid
+     * @param fee       Xcall fee, denominated in ETH
+     * @param toll      Toll taken by this contract, denominated in ETH
+     * @param amtOMNI   Amount of OMNI swapped for
+     */
+    event FillUp(address indexed recipient, uint256 owed, uint256 amtETH, uint256 fee, uint256 toll, uint256 amtOMNI);
+
+    /// @notice Gas limit passed to OmniGasStation.settleUp xcall
+    uint64 internal constant SETTLE_GAS = 100_000;
+
+    /// @notice Denominator for percentage calculations (1000 = 100%)
+    uint256 internal constant PCT_DENOM = 1000;
+
+    /// @notice Address of OmniGasStation on Omni
+    address public gasStation;
+
+    /// @notice Max amt (in native token) that can be swapped in a single tx
+    uint256 public maxSwap;
+
+    /// @notice Percentage toll taken by this contract for each swap, to disencentivize spamming
+    uint256 public toll;
+
+    /// @notice Map recipient to total owed (sum of historical fillUps()), denominated in OMNI.
+    mapping(address => uint256) public owed;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    struct InitParams {
+        address gasStation;
+        address portal;
+        address owner;
+        uint256 maxSwap;
+        uint256 pctCut;
+    }
+
+    function initialize(InitParams calldata p) external initializer {
+        _setGasStation(p.gasStation);
+        _setMaxSwap(p.maxSwap);
+        _setToll(p.pctCut);
+        __XApp_init(p.portal, ConfLevel.Latest);
+        __Ownable_init(p.owner);
+    }
+
+    /**
+     * @notice Swaps msg.value ETH for OMNI and sends it to `recipient` on Omni.
+     *
+     *      Takes an xcall fee and a pct cut. Cut taken to disencentivize spamming.
+     *      Returns the amount of OMNI swapped for.
+     *
+     *      To retry (if OmniGasStation transfer fails), call swap() again with the
+     *      same `recipient`, and msg.value == swapFee().
+     *
+     * @param recipient Address on Omni to send OMNI to
+     */
+    function fillUp(address recipient) public payable whenNotPaused returns (uint256) {
+        // take xcall fee
+        uint256 f = xfee();
+        require(msg.value >= f, "OmniGasPump: insufficient fee");
```

Comment:
can be strictly greater than, because if it's equal, they receive nothing

Reply:
no we want to allow retry, where `msg.value == fee`